### PR TITLE
Budget_app

### PR DESCRIPTION
**Pull request by JakariaJishan**
presentation - https://drive.google.com/file/d/1evbLPkDqLs8VVOE-OF6pT2-hkqDNMNCJ/view?usp=sharing

#### Design
- You should follow these [design guidelines](https://www.behance.net/gallery/19759151/Snapscan-iOs-design-and-branding?tracking_source=), including:
  - Colors.
  - Typography: font face, size and weight.
  - Layout: composition and space between elements.

> NOTE: In these design guidelines there are several UIs that you won't need for this exercise; also, some pages are not given a design and you will create them following the design guidelines of the other pages.

Original design idea by [Gregoire Vella on Behance](https://www.behance.net/gregoirevella).

The [Creative Commons license of the design](https://creativecommons.org/licenses/by-nc/4.0/) requires that you give appropriate credit to the author. Therefore, you must do it in the README of your project.

#### Interactions
- Splash screen
  - A simple page with the name of your app (yes, you need to choose one), and links to the sign up and log in pages.

- Sign up and log in pages
  - The user should be able to register in the app with full name, email and password (all mandatory).
  - The user can log into the app using email and password.
  - If the user is not logged in, they can't access pages that require the user to be logged in (all the pages described below).

- Home page (categories page)
  - When the user logs in, they are presented with the categories page.
  - For each category, the user can see their name, icon and the total amount of all the transactions that belongs to that category.
  - When the user clicks (or taps) on a category item, the application navigates to the transactions page for that category.
  - There is a button "add a new category" at the bottom that brings the user to the page to create a new category.

- Transactions page
  - For a given category, the list of transactions is presented, ordered by the most recent.
  - At the top of the page the user could see the total amount for the category (sum of all of the amounts of the transactions in that category).
  - There is a button "add a new transaction" at the bottom that brings the user to the page to create a new transaction.
  - When the user clicks on the "Back" button (<), the user navigates to the home page.

- "Add a new category" page
  - The user fills out a form to create a new category, indicating their name and icon (both mandatory).
  - The user clicks (or taps) the "Save" button to create the new category, and is taken to the home page on success.
  - When the user clicks on the "Back" button (<), the user navigates to the home page.

- "Add a new transaction" page
  - The user fills out a form to create a new transaction with:
    - name (mandatory)
    - amount (mandatory)
    - categories (mandatory at least one)
  - The user click (or taps) the "Save" button to create the new transaction, and is taken to the transactions page for that category.
  - When the user clicks on the "Back" button (<), the user navigates to the transactions page for that category.

#### Testing requirements
- Create unit and integration tests for all the most important components of your RoR application.

### CODE REVIEW

**Review by BenMukebo — COMMENTED** [review]
## STATUS: INVALID ❌ 

Hi  @JakariaJishan,

I can't miss felicitating your efforts and energy, you made a great effort in this project, unfortunately, I cannot proceed to review your code.

## Invalid Code Review Request 

- [ ] You have submitted a project, that did not follow the GitHub flow.
- Since this is a solo Capston Project, The Pull Request should be made from the “feature branch” into the `dev` branch and not the "dev" to the "main". I suggest you change this. 👍🏻 

Your Code Review Request will be marked as invalid in your Dashboard, so please submit a new one once you are ready 🙏

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me (@BenMukebo) in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

--------

_Invalid Code Review Request **Does not count** into the code reviews limit.

**Comment by JakariaJishan** [comment]
Hi @BenMukebo , thank you for your Review. However, I disagree with your invalidation, here is a screenshot of how to correctly follow [Giflow workflow](https://www.atlassian.com/git/tutorials/comparing-workflows/gitflow-workflow#:~:text=What%20is%20Gitflow%3F,lived%20branches%20and%20larger%20commits.), as you can see below, every feature should reside in its own branch. 👇🏽

image

This is the definition of Feature:

In software development, a feature refers to a distinct and self-contained functionality or characteristic of a software system or application. It represents a specific task or capability that provides value to the users or fulfills a specific requirement.

A complete full-stack project including Database, UI, and all the logic for the actions is clearly not a single Feature but involves several Features, each of those Features needs to be developed in a separated feature-branch according to [Gitflow](https://www.atlassian.com/git/tutorials/comparing-workflows/gitflow-workflow#:~:text=What%20is%20Gitflow%3F,lived%20branches%20and%20larger%20commits.).

> ## STATUS: INVALID ❌
> Hi @JakariaJishan,
> 
> I can't miss felicitating your efforts and energy, you made a great effort in this project, unfortunately, I cannot proceed to review your code.
> 
> ## Invalid Code Review Request
> * [ ]  You have submitted a project, that did not follow the GitHub flow.
> * Since this is a solo Capston Project, The Pull Request should be made from the “feature branch” into the `dev` branch and not the "dev" to the "main". I suggest you change this. 👍🏻
> 
> Your Code Review Request will be marked as invalid in your Dashboard, so please submit a new one once you are ready 🙏
> 
> Cheers and Happy coding!👏👏👏
> 
> Feel free to leave any questions or comments in the PR thread if something is not 100% clear. Please, remember to tag me (@BenMukebo) in your question so I can receive the notification.
> 
> **_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**
> 
> _Invalid Code Review Request **Does not count** into the code reviews limit.

File: app/views/devise/registrations/new.html.erb [hunk 1]
```diff
@@ -0,0 +1,35 @@
+
+<h2 class="text-2xl text-center font-bold my-4">Sign Up</h2>
+
+<%= render 'shared/notice'%>
+<div class="flex flex-col justify-center p-4 bg-white">
+  <%= form_for(resource, as: resource_name, url: registration_path(resource_name)) do |f| %>
```
Comment (KDlamini) [inline app/views/devise/registrations/new.html.erb]: In this milestone, you are required to replicate the EXACT [Behance design guidelines](https://www.behance.net/gallery/19759151/Snapscan-iOs-design-and-branding?tracking_source=) for the registration page.

- There should be a navbar with a background color `#3778c2` that has a back arrow, a heading "REGISTER", and a "next" button.
- The back arrow should navigate back to the home screen, the "next" button should register a new user.
- The upload picture, full name, email, and password fields should be separated by a border with a light grey color `#EDEDEE`.
- The should be no input labels but placeholders, and no border-left and border-right per input field.

image

Your registration form:

image

File: app/views/devise/sessions/new.html.erb [hunk 2]
```diff
@@ -0,0 +1,29 @@
+<h2 class="text-2xl text-center font-bold my-4">Log in</h2>
+
+<%= render 'shared/notice'%>
+<div class="flex flex-col justify-center p-4 bg-white">
+  <%= form_for(resource, as: resource_name, url: session_path(resource_name), html: { class: "mb-4" }) do |f| %>
+    <div class="mb-4">
+      <%= f.label :email, class: "block text-gray-700 text-sm font-bold mb-2" %>
+      <%= f.email_field :email, autofocus: true, autocomplete: "email", class: "shadow appearance-none border rounded w-full py-2 px-3 text-gray-700 leading-tight focus:outline-none focus:shadow-outline", placeholder: "Your Email" %>
```
Comment (KDlamini) [inline app/views/devise/sessions/new.html.erb]: The same thing applies to the login page. Please make sure that your login is an exact match to the [Behance design](https://www.behance.net/gallery/19759151/Snapscan-iOs-design-and-branding?tracking_source=).

Behance design:

image

Your design:

image

File: app/views/categories/index.html.erb [hunk 3]
```diff
@@ -0,0 +1,33 @@
+<div class="flex flex-col justify-center  h-full bg-white">
+  <div class="bg-[#3778c2] flex justify-between items-center text-white p-4 py-6">
+      <% if user_signed_in? %>
+       <%= button_to destroy_user_session_path, method: :delete do %>
+          <%= image_tag("https://img.icons8.com/ios/50/import.png", alt: "exit", width: 20, height: 20) %>
+        <% end %>
+      <% end %>
```
Comment (KDlamini) [inline app/views/categories/index.html.erb:7]: Well done on the categories page. Categories are created following project requirements and design guidelines!!! 👏🏾 💯 

image

File: app/views/categories/show.html.erb [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+       <%= link_to categories_path do %>
+          <%= image_tag("https://img.icons8.com/ios/50/circled-chevron-left.png", alt: "exit", width: 20, height: 20) %>
+        <% end %>
+      <h3> Transections </h3>
+      <h3> <%= button_to category_path(@category), method: :delete, data:{confirm: "Are you sure you want to delete item?"}, class:" opacity-50 hover:opacity-100" do %>
+           <%= image_tag("https://img.icons8.com/ios-glyphs/30/000000/filled-trash.png", alt: "exit", width: 20, height: 20) %>
+          <% end %> </h3>
+  </div>
+    <div class="flex justify-between items-center  border-b border-slate-300 p-4 py-2">
+    <div class="flex items-center gap-px">
```
Comment (KDlamini) [inline app/views/categories/show.html.erb:12]: Well done on your transactions page. Design guidelines are followed, and new transaction form follows project requirements!!! 👏🏾💯

image

File: spec/features/category_spec.rb
Comment (KDlamini) [inline spec/features/category_spec.rb:1]: Great job with the unit and integration tests!!! 💪🏾

image

File: README.md [hunk 5]
```diff
@@ -1 +1,168 @@
...
+- [ ] **Awesome UI**
+
+<p align="right">(<a href="#readme-top">back to top</a>)</p>
+
+## 🤝 Contributing <a name="contributing"></a>
+
+Contributions, issues, and feature requests are welcome!
+
+Feel free to check the [issues page](https://github.com/JakariaJishan/Blog-App/issues).
+
```
Comment (KDlamini) [inline README.md]: The link to your `issues` page is broken, kindly fix it to make your README more professional.  The repo link used is for a blog-app that does not exist.

**Review by KDlamini — CHANGES_REQUESTED** [review]
### Hi @JakariaJishan 🙋‍♀️,  

Great work so far! 💯 
You have done a great job creating a working Budget app using Rails and Devise authentication. 🎊 🥳 
There are some issues that you still need to work on improving your project according to the requirements. You are almost there! 💪🏾 

## HIGHLIGHTS 🌟 

✔️ Good UI implementation
✔️ Good PR title and description
✔️ Models, controllers, migrations, and routes set up correctly
✔️ Database structure follows ERD architecture
✔️ Project is deployed successfully

## Video Presentation 🎥 

✔️  There is a recorded presentation video presenting the features of the project with a maximum length of 5 minutes.
✔️  The video contains a clear and objective description of the project in professional English. The topic and purpose of the project are made clear.
✔️  The video contains a demo of the project, highlighting all the project features.
✖️  The video contains a highlight about a piece of the code the student finds interesting or is proud of making.

## Required Changes ♻️
 
**Kindly look through the comments below 👇 for required changes**

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better._

Cheers 🥂  and Happy coding!!! 👯 
Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me @KDlamini  in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by victorokoroji — APPROVED** [review]
Hi @JakariaJishan, 

# ✔️   Status: APPROVED 🏆 🎉 🎊 

You have done a great job on this project, keep it up 👍 🥇. Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
congratulations!  🎉💯🌟 

## Top highlight 🎯

- Request changes implemented✅ 
 


## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**